### ISSUE TITLE
Checkpointer Feature Enhancements

### ISSUE THREAD

**Issue by fegin** [pytorch/torchtitan#1262]
This document tracks and describes the essential checkpointing features still to be added to TorchTitan.

- [ ] **Full `state_dict` saving**  
  - Support exporting the complete (unsharded) model `state_dict`; many existing formats only handle full `state_dict`.
  - https://github.com/pytorch/torchtitan/pull/1219 is WIP to support this.
  - Need removing FP8 tensor subclass from the `state_dict`.

- [x] **Model `state_dict` mapping**  
  - Provide an interface for users/developers to plug in custom converters between TorchTitan’s `state_dict`/model definitions and other model definitions (e.g., Hugging Face models).

- [x] **Hugging Face format saving**  
  - Depends on full `state_dict` export  
  - Optionally leverages the `model state_dict mapping` interface for users who require conversion
  - Uses the Hugging Face API for saving

- [x] **Hugging Face format loading**  
  - Depends on the `model state_dict interface` as most use cases require conversion from other model definitions  
  - DCP already supports HF loading but needs tighter API integration and performance tuning (collaboration with DCP)

- [ ] **Enhanced checkpoint debugging & comparison tools**  
  - Provide APIs (e.g., per-tensor checksums or diff reports) to pinpoint mismatches in model state, optimizer state, etc.  
  - Streamline root-cause analysis when loaded checkpoints lead to unexpected accuracy changes

- [x] **Complete unit tests**
  - Checkpointer has a lot of logic and branches. We can verify Checkpointer through Mock without using GPUs.

- [ ] **Decouple `state_dict` staging from checkpointing/DCP calls**  
  - Allow staging of the `state_dict` to CPU (or other targets) independently of DCP 
  - Enables downstream workflows (e.g., RL trainers or parameter servers) to consume staged state without invoking DCP

- [ ] **Remove the call to get_model_state_dict and get_optimizer_state_dict**
  - While this originally is viewed as a BE project to demonstrate how to directly get model and optimizer state_dict with canonical FQNs, https://github.com/pytorch/torchtitan/pull/1280 actually depends on this enhancement.

**Comment by fegin** [comment]
cc., @wwwjn update the issue to add one more item

**Comment by gnadathur** [comment]
cc: @saumishr

**Comment by vadimkantorov** [comment]
Maybe adding showcasing/recipe of direct checkpointing to `s3://` via https://github.com/awslabs/s3-connector-for-pytorch (or other native support)?

Also, then there is a question of recommended way of managing/pruning checkpoints during training (keeping checkpoints at regular times + a few latest checkpoints + a few best checkpoints) - especially those saved to s3

**Comment by ivy-zhou** [comment]
Hey @fegin, do you think that DCPv2 already covers the full-state-dict saving case? I looked at the original issue https://github.com/pytorch/torchtitan/issues/1177 and it seems that the user was asking to be able to load .pt files from the checkpoint, which DCPv2 saves by default. Granted there's often duplicate tensors to look through but we also have a CheckpointInspector debugging library we're planning on OSS'ing.

**Comment by fegin** [comment]
We can import and export HF checkpoints. So I think that's good enough for most users.
